error.go

```diff
@@ -20,24 +24,110 @@ import "encoding/json"
 //     }
 //
 type Error struct {
-	Desc string `json:"desc"`
-	Docs string `json:"docs"`
-	Kind string `json:"kind"`
+	Cause error   `json:"cause,omitempty"`
+	Desc  string  `json:"desc,omitempty"`
+	Docs  string  `json:"docs,omitempty"`
+	Kind  string  `json:"kind,omitempty"`
+	Stack []Stack `json:"stack,omitempty"`
 }
 
-func (e *Error) Error() string {
-	return toStringCase(e.Kind)
+type Stack struct {
+	File string `json:"file,omitempty"`
+	Line int    `json:"line,omitempty"`
 }
 
-func (e *Error) GoString() string {
-	return e.String()
+func (e *Error) Error() string {
+	if e.Cause != nil {
+		return e.Cause.Error()
+	}
+
+	k := toStringCase(e.Kind)
+
+	if e.Desc != "" {
+		return fmt.Sprintf("%s: %s", k, e.Desc)
+	}
+
+	return k
 }
 
-func (e *Error) String() string {
-	b, err := json.Marshal(e)
+func (e *Error) MarshalJSON() ([]byte, error) {
+	type ErrorClone Error
+
+	b, err := json.Marshal(&struct {
+		*ErrorClone
+	}{
+		ErrorClone: (*ErrorClone)(e),
+	})
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
 
-	return string(b)
+	return b, nil
+}
+
+// SetDescf fills the error's Desc property at runtime. This is useful when
+// returning the route cause of an error at runtime. We usually reuse certain
+// error types for different situations. Then contextual information need to be
+// transported to make the error more useful. The example below shows how this
+// could look like.
+//
+//     return microerror.Mask(executionFailedError.SetDescf("NS record %#q for HostedZone %#q not found", name, id))
+//
+// Note that modifying the description of a received error is considered an
+// anti-pattern. This is not easily possible since the usually used error
+// interface requires type assertions for *Error in order to change its
+// description again. When receiving errors one should only use Mask and forward
+// the received error to the next caller. When extending the description of one
+// error with the message of another and in case the error kind is the same for
+// both of these errors, SetDescf removes the prefix in f equal to Kind of e. If
+// we would not prune the prefix, we would have issues with repetitive error
+// kinds reflected in resulting error messages. One famous example looks as
+// follows.
+//
+//     scaling.min and scaling.max must be equal on azure: invalid request error: invalid request error: invalid request error
+//
+func (e *Error) SetDescf(f string, args ...interface{}) *Error {
```

Comment:
I dislike this part very much. API is cumbersome compared to previous `Maskf`. It also mutates the error which we define as a pointer. I have some ideas in my head. Let's pair on that and talk.